[mlir][vector] Add support for `vector.multi_reduction` and `vector.shape_cast` distribution.

charithaintc opened the pull request:
This PR adds support for `vector.multi_reduction` distribution. Currently only 2D to 1D reductions are supported (col/row reductions) and assumes the inner dimension of source vector is distributed among lanes (each lane owns columns of the source vector).

* **Col Reduce** : Each lane owns the data. Each lane can do its own `vector.reduction` of the column data.
* **Row Reduce**: Requires shuffling data with neighbors. Simply rewrite the `multi_reduction` in terms of `reduction` ops.

PR also include changes in `vector.shape_cast` distribution to consider the distributed type of shape case source if given by a `DistributionMapFn`

charithaintc committed:
save

charithaintc committed:
save

charithaintc committed:
Merge branch 'main' into vector_multi_reduction_distr

charithaintc commented:
cc @Garra1980

charithaintc committed:
save

charithaintc committed:
save

charithaintc committed:
save

charithaintc committed:
save

adam-smnk reviewed (commented):
General direction looks good to me.
I'll let others have a closer look at the distribution logic.

charithaintc committed:
Merge branch 'main' into vector_multi_reduction_distr

charithaintc committed:
save

charithaintc commented on mlir/lib/Dialect/Vector/Transforms/VectorDistribute.cpp:
```diff
@@ -1996,6 +2027,113 @@ struct WarpOpReduction : public WarpDistributionPattern {
+struct VectorMultiDimReductionDistribution : public WarpDistributionPattern {
+  VectorMultiDimReductionDistribution(MLIRContext *context,
+                                      PatternBenefit benefit = 1)
+      : WarpDistributionPattern(context, benefit) {}
```
fixed.

charithaintc committed:
save

charithaintc commented:
> General direction looks good to me. I'll let others have a closer look at the distribution logic.

Hi @adam-smnk, Thanks for the reviews. I addressed the concerns. Please let me know if you have any additional concerns.

I also properly documented the restrictions in this version.

chencha3 commented on mlir/lib/Dialect/Vector/Transforms/VectorDistribute.cpp:
```diff
@@ -977,44 +977,75 @@ struct WarpOpBroadcast : public WarpDistributionPattern {
+    // distributed source type according to following rules:
+    // 1. If the source type is yielded from the warp op, we can use the
+    //    matching warp result type as the distributed source type.
+    // 2. If the source type is not yielded from the warp op, we need
```
Could you give some example. Will it happen that there are conflicts between these two rules?

charithaintc replied:
I added two examples for the 2 cases as comment.

For both row and col reduction we assume that cols of the source vector is owned by each lane (i.e. in xegpu layouts this will be [1, 16]). Based on that we handle the reduction logic. 
Given this layout,
Col reduction is easy : just reduce your own data.
Row reduction: needs to shuffle data with neighbors and do a tree like reduce (aka butterfly reduction with shuffles)

however source layout can also be [16, 1]. This case is not supported because the vector distribution infra does not allow me to express such layout currently (it always start distributing the vector from innermost dim).  I am working on some proposal to improve this.

chencha3 commented on mlir/lib/Dialect/Vector/Transforms/VectorDistribute.cpp:
```diff
@@ -1996,6 +2027,134 @@ struct WarpOpReduction : public WarpDistributionPattern {
+    // case each lane owns its portion of the result (i.e. result is also
+    // distributed).
+    // 3. If reduction dim == 1, its a row reduction that require cross lanes
+    // shuffles. In this case result is not distributed and broadcasted instead.
```
not distributed but broadcasted instead?

charithaintc replied:
yes. for row reductions each lane owns the whole reduced result. Example.
```
%r = vector_multi_reduction <add> %src, %acc, [1] : vector<8x16xf32> -> vector<8xf32>
```
In this case each lane own a col of 8x16xf32 (i.e. 8x1xf32). Each lane will do a shuffle and add reduce log2(16) times to get the final result. In this case each lane will have the final result of 8x1xf32 at the end. meaning that the result is eventually broadcasted to each lane.

charithaintc replied:
I see. thanks for clarifying. I fixed it.

chencha3 commented on mlir/lib/Dialect/Vector/Transforms/VectorDistribute.cpp:
```diff
@@ -1996,6 +2027,134 @@ struct WarpOpReduction : public WarpDistributionPattern {
+                                               reductionOp.getSource(), i);
+      Value acc = vector::ExtractOp::create(rewriter, reductionOp.getLoc(),
+                                            reductionOp.getAcc(), i);
+      Value rowReduce = vector::ReductionOp::create(
```
I am curious how is ShuffleOp inserted?

chencha3 reviewed (commented):
Overall, it looks good to me. It might be helpful to have someone familiar with the distribution mechanism review it for accuracy.

charithaintc committed:
save work

charithaintc committed:
save work

adam-smnk commented on mlir/test/Dialect/Vector/vector-warp-distribute.mlir:
```diff
@@ -850,6 +850,85 @@ func.func @vector_reduction_acc(%laneid: index) -> (f32) {
+func.func @vector_multi_reduction_row_reduce(%laneid: index) -> vector<2xf32> {
+  %zero = arith.constant dense<0.0> : vector<2xf32>
+  %r = gpu.warp_execute_on_lane_0(%laneid)[32] -> (vector<2xf32>) {
+    %0 = "some_def"() : () -> (vector<2x32xf32>)
```
nit: maybe use larger innermost dim such that canonicalization doesn't remove single element `vector.reduction`?

adam-smnk reviewed (commented):
Overall seems fine. Minor comments.

@chencha3 I'm not sure about shuffles, if you think they're fine then PR should be good to go

charithaintc committed:
Merge branch 'main' into vector_multi_reduction_distr

charithaintc commented on mlir/lib/Dialect/Vector/Transforms/VectorDistribute.cpp:
```diff
@@ -1020,44 +1020,75 @@ struct WarpOpBroadcast : public WarpDistributionPattern {
+    bool isResultDistributed = distributedResultType.getNumElements() <
+                               oldCastOp.getResultVectorType().getNumElements();
+
+    // If the result is not distributed, source distribted type is the same
```
fixed.

charithaintc committed:
save work

Jianhui-Li commented on mlir/test/Dialect/Vector/vector-warp-distribute.mlir:
```diff
@@ -1567,6 +1646,40 @@ func.func @warp_propagate_shape_cast(%laneid: index, %src: memref<32x4x32xf32>)
+
+// CHECK-PROP-LABEL: func.func @warp_propagate_shape_cast_2d_to_2d
+// CHECK-PROP:    %[[READ:.*]] = vector.transfer_read {{.*}} {in_bounds = [false, true]} : memref<64x32xf32>, vector<2x32xf32>
+// CHECK-PROP:    %[[CAST:.*]] = vector.shape_cast %[[READ]] : vector<2x32xf32> to vector<32x2xf32>
```
why it is not 
 %[[CAST:.*]] = vector.shape_cast %[[READ]] : vector<**64x1**xf32> to vector<32x2xf32>

charithaintc replied:
the inner dim of the transfer_read result is distributed in this case. so it becomes 64x32 -> 2x32 after distribution.
